slowquant/unitary_coupled_cluster/optimizers.py

```diff
@@ -0,0 +1,312 @@
+                callback=print_progress,
+            )
+            res = optimizer.minimize(self.fun, x0)
+
+        else:
+            raise ValueError(f"Got an unkonwn optimizer {self.method}")
+        result = Result()
+        result.x = res.x
+        result.fun = res.fun
+        return result
+
+
+class RotoSolve:
+    r"""Rotosolve optimizer.
+
+    Implemenation of Rotosolver assuming three eigenvalues for generators.
+    This works for fermionic generators of the type:
+
+    .. math::
+        \hat{G}_{pq} = \hat{a}^\dagger_p \hat{a}_q - \hat{a}_q^\dagger \hat{a}_p
+
+    and,
+
+    .. math::
+        \hat{G}_{pqrs} = \hat{a}^\dagger_p \hat{a}^\dagger_q \hat{a}_r \hat{a}_s - \hat{a}^\dagger_s \hat{a}^\dagger_r \hat{a}_p \hat{a}_q
+
+    Rotosolve works by exactly reconstrucing the energy function in a single parameter:
+
+    .. math::
+        E(x) = \frac{\sin\left(\frac{2R+1}{2}x\right)}{2R+1}\sum_{\mu=-R}^{R}E(x_\mu)\frac{(-1)^\mu}{\sin\left(\frac{x - x_\mu}{2}\right)}
+
+    With :math:`R` being the number of different positive differences between eigenvalues, and :math:`x_\mu=\frac{2\mu}{2R+1}\pi`.
+
+    After the function :math:`E(x)` have been reconstruced the global minima of the function can be found classically.
+
+    #. 10.22331/q-2021-01-28-391, Algorithm 1
+    #. 10.22331/q-2022-03-30-677, Eq. (57)
+    """
+
+    def __init__(
+        self,
+        R: dict[str, int],
+        param_names: Sequence[str],
+        maxiter: int = 30,
+        tol: float = 1e-6,
+        callback: Callable[[list[float]], None] | None = None,
+    ) -> None:
+        """Initialize Rotosolver.
+
+        Args:
+            R: R parameter used for the function reconstruction.
+            param_names: Names of parameters, used to index R.
+            maxiter: Maximum number of iterations (sweeps).
+            tol: Convergence tolerence.
+            callback: Callback function, takes only x (parameters) as an argument.
+        """
+        self._callback = callback
+        self.max_iterations = maxiter
+        self.threshold = tol
+        self.max_fail = 3
+        self._R = R
+        self._param_names = param_names
+
+    def minimize(self, f: Callable[[list[float]], float | list[float]], x0: Sequence[float]) -> Result:
+        """Run minimization.
+
+        Args:
+            f: Function to be minimzed, can only take one argument.
+            x: Changable parameters of f.
+
+        Returns:
+            Minimization results.
+        """
+        f_best = float(10**20)
+        x = list(x0).copy()
+        x_best = x.copy()
+        fails = 0
+        res = Result()
+        for _ in range(self.max_iterations):
+            for i, par_name in enumerate(self._param_names):
+                e_vals = get_energy_evals(f, x, i, self._R[par_name])
+                f_reconstructed = partial(reconstructed_f, energy_vals=e_vals, R=self._R[par_name])
+                values = f_reconstructed(np.linspace(-np.pi, np.pi, int(1e4)))
```

Comment:
Could you maybe just add a few comments here. This would be really nice. Like 1 line comments for each of these important steps

Reply:
I have granted you some comments ...